### PR TITLE
Add NGINX integration with 3scale on Kubernetes

### PR DESCRIPTION
**Pull request by krish7919**

### CODE REVIEW

**Comment by krish7919** [comment]
One of the things missing from this PR is the documentation.
The documentation for this PR is still in PR #1305.
I will hold on to it for now once I have the complete docs ready for deploying everything.
Else we can leave this open till as such time the docs shape up.

**Comment by ttmc** [comment]
I see our work on Kubernetes deployment as a work in progress (and said as much in the change log). It doesn't have to be fully-functional or fully-documented for the version 0.10.0 release next week. Mainly we need the single-node case operational so that REDACTED can use it.

File: k8s/nginx-3scale/nginx-3scale-dep.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    spec:
+      terminationGracePeriodSeconds: 10
+      containers:
+      - name: nginx-3scale
+        image: bigchaindb/nginx_3scale:0.1
+        # TODO(Krish): Change later to IfNotPresent
+        imagePullPolicy: Always
+        env:
+        - name: MONGODB_FRONTEND_PORT
+          value: $(NGX_INSTANCE_0_SERVICE_PORT_NGX_PUBLIC_MDB_PORT)
```
Comment (ttmc) [inline k8s/nginx-3scale/nginx-3scale-dep.yaml]: What does the `$(STUFF)` syntax mean?

File: k8s/nginx-3scale/nginx-3scale-secret.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Certificate data should be base64 encoded before embedding them here by using
+# `cat cert.pem | base64 -w 0 > cert.pem.b64` and then copy the resulting
+# value here. Same goes for cert.key.
+# Ref: https://kubernetes.io/docs/concepts/configuration/secret/
```
Comment (ttmc) [inline k8s/nginx-3scale/nginx-3scale-secret.yaml]: Thanks for including these comments here.

File: k8s/nginx-3scale/nginx-3scale-svc.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: ngx-instance-0
+  namespace: default
+  labels:
+    name: ngx-instance-0
+  annotations:
+    # NOTE: the following annotation is a beta feature and
+    # only available in GCE/GKE and Azure as of now
```
Comment (ttmc) [inline k8s/nginx-3scale/nginx-3scale-svc.yaml]: Maybe include a link to an explanation in the comments as well, e.g.

https://kubernetes.io/docs/tutorials/services/source-ip/

File: k8s/nginx-3scale/nginx-3scale-dep.yaml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          protocol: TCP
+        volumeMounts:
+        - name: https
+          mountPath: /usr/local/openresty/nginx/conf/ssl/
+          readOnly: true
+        resources:
+          limits:
+            cpu: 200m
+            memory: 768Mi
+        livenessProbe:
```
Comment (ttmc) [inline k8s/nginx-3scale/nginx-3scale-dep.yaml]: No readinessProbe?

File: k8s/nginx-3scale/nginx-3scale-dep.yaml (hunk 1 above)
Comment (krish7919) [inline k8s/nginx-3scale/nginx-3scale-dep.yaml]: In Kubernetes config files, that's how to you refer the env variables.
This basically states: Set the env var `MONGODB_FRONTEND_PORT` to the value defined by the env var `NGX_INSTANCE_0_SERVICE_PORT_NGX_PUBLIC_MDB_PORT`.

File: k8s/nginx-3scale/nginx-3scale-dep.yaml (hunk 4 above)
Comment (krish7919) [inline k8s/nginx-3scale/nginx-3scale-dep.yaml]: It's hard to define nginx being `ready` in this case.
Should it only be when the health-check port is up?
Should it be when MongoDB frontend port and/or BigchainDB backend port is up?
Also, these ports are publicly visible so if the services in the backend fails, then the proxy will stop forwarding packets.

**Comment by ttmc** [comment]
:+1:

**Pull request merged**